lab/skel_graph/astar/main.cpp

```diff
@@ -0,0 +1,216 @@
+    std::vector<int> path;
+    std::pair<int, int> source;
+    std::pair<int, int> destination;
+    Matrix adj_matrix;
+    std::set<std::pair<int, int>> closed;
+    std::set<std::pair<int, int>> open;
+    Parents parent;
+
+    // matrix of a tuple <f,g, h> for every value
+    std::vector<std::vector<std::tuple<double, double, double>>> fgh;
+
+    void read_input() {
+        std::cin >> n >> m;
+        adj_matrix = Matrix(n + 1, std::vector<int> (m + 1 , 0));
+        parent = Parents(n + 1, std::vector<std::pair<int, int>>
+        (m + 1, {-1, -1}));
+        fgh =  std::vector<std::vector<std::tuple<double, double, double>>>
+        (n + 1, std::vector<std::tuple<double, double, double>>
+        (m + 1, std::make_tuple(DBL_MAX, DBL_MAX, DBL_MAX)));
+
+        for (int i = 1; i <= n; i++) {
+            for (int j = 1; j <= m; j++) {
+                std::cin >> adj_matrix[i][j];
+            }
+        }
+
+
+        std::cin >> source.first >> source.second;
+        std::cin >> destination.first >> destination.second;
+    }
+
+
+    void print_output() {
+        for (auto &pos : buildPath(destination)) {
+            std::cout<< "Trec prin punctul" << pos.first << " " << pos.second;
+            std::cout << "\n";
+        }
+    }
+
+    std::vector<std::pair <int, int>>  expand(std::pair <int, int> node) {
+        std::vector<std::pair <int, int>> directions =
+        {{1, 1}, {1, 0}, {1, -1}, {0, 1}, {0, -1}, {-1, 1}, {-1, 0}, {-1, -1}};
+        std::vector<std::pair <int, int>> succ;
+        for (auto const &direction : directions) {
+            int dx = direction.first;
+            int dy = direction.second;
+            int x = node.first + dx;
+            int y = node.second + dy;
+            if (x >= 1 && x <= n && y >= 1 && y <= n && adj_matrix[x][y] == 1) {
+                succ.push_back({x, y});
+            }
+        }
+        return succ;
+    }
+
+    bool is_goal(std::pair<int, int> node) {
+        return (node == destination);
+    }
+
+    double euclidian_distance_heuristic(std::pair<int, int> node) {
+        int abs_x = abs(node.first - destination.first);
+        int abs_y = abs(node.second - destination.second);
+        return sqrt(abs_x * abs_x + abs_y * abs_y);
+    }
+
+    std::vector<std::pair <int, int>> buildPath(std::pair<int, int> node) {
+        auto current = node;
+        std::vector<std::pair<int, int>> path;
+        while (!(current.first == -1 && current.second == -1)) {
+            path.push_back(current);
+            current = parent[current.first][current.second];
+        }
+        std::reverse(path.begin(), path.end());
+        return path;
+    }
+
+    bool get_result() {
+        // initializari
+        open.clear();  // multimea nodurilor in curs de explorare
+        closed.clear();  // multimea nodurilor expandate
+
+        // functie lambda pentru a compara doua perechi
+        auto comparator = []
+        (std::pair<double, std::pair<int, int>> x1,
```

Comment:
This is ugly!

What you need is auto

[](const auto& x1, const auto& x2){ ... }